src/main/AllocationTable.cpp

```diff
@@ -0,0 +1,286 @@
+      if (bucket.isFirst()) {
+        if (bucket.isSuperblock()) {
+          // This is a large superblock.
+          AllocatorSuperblock* superblock = bucket.asSuperblock();
+          superblock->applyToBlocks(fun);
+        } else {
+#ifdef ALLOCATION_CHECK
+          // Extra consistency check.
+          if (!GCNodeAllocator::lookupPointer(bucket.asPointer())) {
+            allocerr("apply to all blocks iterating over non-alloc'd pointer");
+          }
+#endif
+          fun(bucket.asPointer());
+        }
+      }
+    }
+  }
+}
+
+void rho::AllocationTable::printSummary() {
+  printf("Allocation Table Summary:\n");
+  int size = 0;
+  for (int i = 0; i < m_num_buckets; i += 16) {
+    int count = 0;
+    for (int j = 0; j < 16 && i + j < m_num_buckets; ++j) {
+      Allocation& bucket = m_buckets[i + j];
+      if (!bucket.isEmpty() && !bucket.isDeleted()) {
+        count += 1;
+      }
+    }
+    if (count) {
+      printf("%X", count);
+    } else {
+      printf(".");
+    }
+    size += count;
+  }
+  printf("\n");
+  printf("table size: %d\n", size);
+}
+
+bool rho::AllocationTable::tryInsert(Allocation* table, uintptr_t pointer, size_t size) {
+  size_t key_end = endKey(pointer, size);
+  bool first = true;
+  for (size_t key = startKey(pointer); key < key_end; ++key) {
+    size_t hash = pointerHash(key << LOW_BITS);
+    bool added = false;
+    for (int i = 0; i < m_num_buckets; ++i) {
+      Allocation& bucket = table[hash];
+      if (bucket.isEmpty() || bucket.isDeleted()) {
+        new (&table[hash])Allocation(reinterpret_cast<void*>(pointer), size, first);
+        first = false;
+        added = true;
+        break;
+      }
+      m_num_insert_collisions += 1;
+      hash = probeFunc(hash, i);
+    }
+    if (!added) {
+      // Failed to insert the current key.
+      return false;
+    }
+  }
+  // All keys inserted.
+  return true;
+}
+
+size_t rho::AllocationTable::pointerHash(uintptr_t pointer) {
+  pointer >>= LOW_BITS;
+  pointer <<= 4;  // Spread out hash keys to leave room for collision buckets.
+  size_t low = pointer & 0xFFFFFFFF;
+  size_t hi = pointer >> 32;
+  size_t hash = low ^ hi;
+  low = hash & 0xFFFF;
+  hi = (hash >> 16) & 0xFFFF;
+  hash = (low >> 16) ^ low ^ hi ^ (pointer & (~0xFFFF));
+  return hash & m_hash_mask;
+}
+
+size_t rho::AllocationTable::probeFunc(size_t prev_hash, int i) {
+  return (prev_hash + i + 1) & m_hash_mask;
+}
+
+size_t rho::AllocationTable::startKey(uintptr_t pointer) {
+  return pointer >> LOW_BITS;
+}
+
+size_t rho::AllocationTable::endKey(uintptr_t pointer, size_t size) {
+  size_t key_end = (pointer + (1 << size) + ((1 << LOW_BITS) - 1)) >> LOW_BITS;
+  if (startKey(pointer) == key_end) {
+    // Ensure that at least one entry is added to the table.
+    // If the pointer was exactly aligned to LOW_BITS then
+    // key_end will be equal to key, otherwise key_end will point
+    // to the next hash bucket past the allocation range.
```

Comment:
Ah, you are right. Since the allocation size is never 0 it always pushes the end key to the next hash key regardless of whether the allocation was aligned, so I can just delete that check!